relayer/workers/beefyrelayer/beefy-ethereum-listener.go

```diff
@@ -181,3 +241,152 @@ func (li *BeefyEthereumListener) processLightClientEvents(ctx context.Context, e
 		li.dbMessages <- updateCmd
 	}
 }
+
+// queryFinalVerificationSuccessfulEvents queries ContractFinalVerificationSuccessful events from the LightClientBridge contract
+func (li *BeefyEthereumListener) queryFinalVerificationSuccessfulEvents(ctx context.Context, start uint64,
+	end *uint64) ([]*lightclientbridge.ContractFinalVerificationSuccessful, error) {
+	var events []*lightclientbridge.ContractFinalVerificationSuccessful
+	filterOps := bind.FilterOpts{Start: start, End: end, Context: ctx}
+
+	iter, err := li.lightClientBridge.FilterFinalVerificationSuccessful(&filterOps)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		more := iter.Next()
+		if !more {
+			err = iter.Error()
+			if err != nil {
+				return nil, err
+			}
+			break
+		}
+
+		events = append(events, iter.Event)
+	}
+
+	return events, nil
+}
+
+// processHistoricalFinalVerificationSuccessfulEvents processes historical FinalVerificationSuccessful
+// events, updating the status of matched BEEFY justifications in the database
+func (li *BeefyEthereumListener) processHistoricalFinalVerificationSuccessfulEvents(ctx context.Context,
+	blockNumber, latestBlockNumber uint64) {
+	// Query previous FinalVerificationSuccessful events and update the status of BEEFY justifications in database
+	events, err := li.queryFinalVerificationSuccessfulEvents(ctx, blockNumber, &latestBlockNumber)
+	if err != nil {
+		li.log.WithError(err).Error("Failure fetching event logs")
+	}
+	li.log.Info(fmt.Sprintf(
+		"Found %d FinalVerificationSuccessful events between blocks %d-%d",
+		len(events), blockNumber, latestBlockNumber),
+	)
+
+	for _, event := range events {
+		// Fetch validation data from contract using event.ID
+		validationData, err := li.lightClientBridge.ContractCaller.ValidationData(nil, event.Id)
+		if err != nil {
+			li.log.WithError(err).Error(fmt.Sprintf("Error querying validation data for ID %d", event.Id))
+		}
+
+		// Attempt to match items in database based on their payload
+		itemFoundInDatabase := false
+		items := li.beefyDB.GetItemsByStatus(store.InitialVerificationTxConfirmed)
+		for _, item := range items {
+			generatedPayload := li.simulatePayloadGeneration(*item)
+			if generatedPayload == validationData.Payload {
+				li.log.Info("Deleting finalized item from the database'")
+				deleteCmd := store.NewDatabaseCmd(item, store.Delete, nil)
+				li.dbMessages <- deleteCmd
+
+				itemFoundInDatabase = true
+				break
+			}
```

Comment:
Do we really need to delete those items? It might be worthwhile to keep them with a `Done` status for audit/debugging purposes?

Reply:
Here we're using [soft delete](https://gorm.io/docs/delete.html#Soft-Delete) by default, which sets the `DeletedAt` field and filters the object out of normal database queries (the item can still be found with special query types).

Reply:
Awesome!